File: crates/traderview-db/src/house_hacking.rs
//! House hacking — buy a small multi-unit, live in one, rent the rest.
//!
//! The canonical wealth-building starter move: the tenants' rent offsets
//! (or erases) your housing cost, so the money that would have been rent
//! goes to principal and savings instead. This nets the rental income
//! against the full carrying cost (P&I + tax + insurance + maintenance +
//! HOA) to show what you actually pay to live there, compares it to
//! renting a comparable place, and reports the property's standalone cash
//! flow once you move out and rent every unit.
//!
//! Pure compute — the P&I uses the shared
//! [`crate::mortgage_amortization::monthly_payment`] so there's one
//! amortization formula in the codebase.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct HouseHackInput {
    pub home_price_usd: f64,
    pub down_payment_usd: f64,
    pub apr_pct: f64,
    pub term_months: u32,
    pub total_units: u32,
    pub owner_units: u32,
    /// Monthly rent collected per rented unit.
    pub rent_per_unit_usd: f64,
    pub monthly_tax_usd: f64,
    pub monthly_insurance_usd: f64,
    pub monthly_maintenance_usd: f64,
    pub monthly_hoa_usd: f64,
    /// What renting a comparable place for yourself would cost — the
    /// baseline the house hack is measured against.
    pub comparable_rent_usd: f64,
}

#[derive(Debug, Clone, Serialize, Default)]
pub struct HouseHackResult {
    pub principal_financed_usd: f64,
    pub monthly_pi_usd: f64,
    pub rented_units: u32,
    pub rental_income_usd: f64,
    /// P&I + tax + insurance + maintenance + HOA.
    pub total_housing_cost_usd: f64,
    /// Carrying cost net of rent — what you actually pay to live there.
    /// Negative means the tenants more than cover the place.
    pub net_housing_cost_usd: f64,
    /// `comparable_rent − net_housing_cost`: monthly saved vs renting.
    pub savings_vs_renting_usd: f64,
    /// Does one unit's rent already cover the mortgage P&I?
    pub rent_covers_pi: bool,
    /// Cash flow once you move out and rent every unit.
    pub full_rental_cash_flow_usd: f64,
}

pub fn compute(i: &HouseHackInput) -> HouseHackResult {
    let principal = (i.home_price_usd - i.down_payment_usd).max(0.0);
    let monthly_pi = crate::mortgage_amortization::monthly_payment(principal, i.apr_pct, i.term_months);

    let owner = i.owner_units.min(i.total_units);
    let rented = i.total_units.saturating_sub(owner);
    let rental_income = rented as f64 * i.rent_per_unit_usd;

    let operating = i.monthly_tax_usd
        + i.monthly_insurance_usd
        + i.monthly_maintenance_usd
        + i.monthly_hoa_usd;
    let total_housing_cost = monthly_pi + operating;
    let net_housing_cost = total_housing_cost - rental_income;

    let full_rental_income = i.total_units as f64 * i.rent_per_unit_usd;

    HouseHackResult {
        principal_financed_usd: principal,
        monthly_pi_usd: monthly_pi,
        rented_units: rented,
        rental_income_usd: rental_income,
        total_housing_cost_usd: total_housing_cost,
        net_housing_cost_usd: net_housing_cost,
        savings_vs_renting_usd: i.comparable_rent_usd - net_housing_cost,
        rent_covers_pi: rental_income >= monthly_pi,
        full_rental_cash_flow_usd: full_rental_income - total_housing_cost,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn duplex() -> HouseHackInput {
        HouseHackInput {
            home_price_usd: 400_000.0,
            down_payment_usd: 80_000.0,
            apr_pct: 6.0,
            term_months: 360,
            total_units: 2,
            owner_units: 1,
            rent_per_unit_usd: 1_500.0,
            monthly_tax_usd: 400.0,
            monthly_insurance_usd: 100.0,
            monthly_maintenance_usd: 200.0,
            monthly_hoa_usd: 0.0,
            comparable_rent_usd: 1_800.0,
        }
    }

    #[test]
    fn nets_rent_against_carrying_cost() {
        let r = compute(&duplex());
        assert!((r.principal_financed_usd - 320_000.0).abs() < 1e-6);
        assert_eq!(r.rented_units, 1);
        assert!((r.rental_income_usd - 1_500.0).abs() < 1e-9);
        // Identities the result must satisfy, independent of the exact P&I.
        assert!((r.total_housing_cost_usd - (r.monthly_pi_usd + 700.0)).abs() < 1e-6);
        assert!((r.net_housing_cost_usd - (r.total_housing_cost_usd - r.rental_income_usd)).abs() < 1e-6);
        assert!((r.savings_vs_renting_usd - (1_800.0 - r.net_housing_cost_usd)).abs() < 1e-6);
    }

    #[test]
    fn monthly_pi_matches_shared_amortization() {
        let r = compute(&duplex());
        let expect = crate::mortgage_amortization::monthly_payment(320_000.0, 6.0, 360);
        assert!((r.monthly_pi_usd - expect).abs() < 1e-9);
        // ~$1,918/mo on a $320k 30-yr at 6%.
        assert!((r.monthly_pi_usd - 1_918.0).abs() < 5.0, "got {}", r.monthly_pi_usd);
        assert!(!r.rent_covers_pi); // one unit's $1,500 < ~$1,918 P&I
    }

    #[test]
    fn fourplex_can_pay_you_to_live_there() {
        // 4 units, live in one, rent three at $1,500 = $4,500 income.
        let r = compute(&HouseHackInput {
            total_units: 4,
            ..duplex()
        });
        assert_eq!(r.rented_units, 3);
        assert!((r.rental_income_usd - 4_500.0).abs() < 1e-9);
        // Income > carrying cost → negative net housing cost (paid to live).
        assert!(r.net_housing_cost_usd < 0.0, "net {}", r.net_housing_cost_usd);
        assert!(r.rent_covers_pi);
    }

    #[test]
    fn full_rental_cash_flow_after_moving_out() {
        let r = compute(&duplex());
        // Both units rented = $3,000 less the total carrying cost.
        let expect = 3_000.0 - r.total_housing_cost_usd;
        assert!((r.full_rental_cash_flow_usd - expect).abs() < 1e-6);
    }

    #[test]
    fn owner_units_capped_at_total() {
        // Claiming to occupy more units than exist leaves zero rented.
        let r = compute(&HouseHackInput {
            total_units: 2,
            owner_units: 5,
            ..duplex()
        });
        assert_eq!(r.rented_units, 0);
        assert_eq!(r.rental_income_usd, 0.0);
    }
}

File: frontend/js/views/house_hacking.js
// House Hacking — buy a small multi-unit, live in one, rent the rest.
// Nets rental income against the carrying cost to show what you actually
// pay to live there, via /calc/house-hacking.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import * as enh from '../calc_enhance.js';

const VIEW = 'house-hacking';
let lastReport = null;
let lastBody = null;

const FIELDS = [
    ['home_price_usd', 'Home price ($)', 400000],
    ['down_payment_usd', 'Down payment ($)', 80000],
    ['apr_pct', 'Mortgage APR (%)', 6],
    ['term_months', 'Term (months)', 360],
    ['total_units', 'Total units', 2],
    ['owner_units', 'Units you occupy', 1],
    ['rent_per_unit_usd', 'Rent per unit ($/mo)', 1500],
    ['monthly_tax_usd', 'Property tax ($/mo)', 400],
    ['monthly_insurance_usd', 'Insurance ($/mo)', 100],
    ['monthly_maintenance_usd', 'Maintenance ($/mo)', 200],
    ['monthly_hoa_usd', 'HOA ($/mo)', 0],
    ['comparable_rent_usd', 'Comparable rent for you ($/mo)', 1800],
];
const INT_FIELDS = new Set(['term_months', 'total_units', 'owner_units']);

const money = (n) => (n < 0 ? '-$' : '$') + Math.abs(Number(n)).toLocaleString(undefined, { maximumFractionDigits: 0 });

export async function renderHouseHacking(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.househack.h1.title">// HOUSE HACKING</span></h1>
        <p class="muted small" data-i18n="view.househack.hint.intro">
            Buy a 2–4 unit property, live in one unit, rent the rest. The tenants' rent
            offsets your carrying cost (mortgage + tax + insurance + maintenance + HOA),
            so the money that would have been rent builds equity instead. This shows what
            you actually pay to live there, how it compares to renting, and the property's
            cash flow once you move out and rent every unit.
        </p>
        <div class="chart-panel">
            <h2 data-i18n="view.househack.h2.inputs">The deal</h2>
            <form id="hh-form" class="inline-form">
                ${FIELDS.map(([key, label, def]) => `
                    <label><span data-i18n="view.househack.label.${key}">${label}</span>
                        <input type="number" step="${INT_FIELDS.has(key) ? '1' : '0.01'}" min="0" name="${key}" value="${def}" required></label>
                `).join('')}
                <button class="primary" type="submit" data-i18n="view.househack.btn.run">Run the numbers</button>
            </form>
            <div id="hh-tools" class="ce-toolbar"></div>
        </div>
        <div id="hh-result"></div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#hh-form');
    enh.prefillForm(form, enh.readHashInputs());
    const readBody = () => {
        const fd = new FormData(form);
        const body = {};
        for (const [key] of FIELDS) body[key] = Number(fd.get(key)) || 0;
        return body;
    };
    enh.mountToolbar(mount.querySelector('#hh-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'house-hacking.csv' });
    form.addEventListener('submit', async (e) => {
        e.preventDefault();
        const body = readBody();
        try {
            const r = await api.calcHouseHacking(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r);
        } catch (err) {
            showToast(err.message || t('view.househack.toast.error'), { level: 'error' });
        }
    });
    form.dispatchEvent(new Event('submit'));
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['monthly_pi_usd', r.monthly_pi_usd],
        ['rental_income_usd', r.rental_income_usd],
        ['total_housing_cost_usd', r.total_housing_cost_usd],
        ['net_housing_cost_usd', r.net_housing_cost_usd],
        ['savings_vs_renting_usd', r.savings_vs_renting_usd],
        ['full_rental_cash_flow_usd', r.full_rental_cash_flow_usd],
    ];
}

function renderResult(mount, r) {
    const el = mount.querySelector('#hh-result');
    const netCls = Number(r.net_housing_cost_usd) <= 0 ? 'pos' : '';
    // Carrying cost vs rental income vs your net housing cost.
    const chart = enh.svgBarChart([
        { label: 'Carry', value: -r.total_housing_cost_usd },
        { label: 'Rent', value: r.rental_income_usd },
        { label: 'Net cost', value: -r.net_housing_cost_usd },
    ]);
    const saveCls = Number(r.savings_vs_renting_usd) >= 0 ? 'pos' : 'neg';
    const cfCls = Number(r.full_rental_cash_flow_usd) >= 0 ? 'pos' : 'neg';
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.househack.h2.result">The numbers</h2>
            <div class="cards">
                <div class="card"><div class="label" data-i18n="view.househack.card.pi">Mortgage P&amp;I</div>
                    <div class="value">${money(r.monthly_pi_usd)}/mo</div></div>
                <div class="card"><div class="label" data-i18n="view.househack.card.income">Rental income (${r.rented_units})</div>
                    <div class="value pos">${money(r.rental_income_usd)}/mo</div></div>
                <div class="card"><div class="label" data-i18n="view.househack.card.carry">Total carrying cost</div>
                    <div class="value">${money(r.total_housing_cost_usd)}/mo</div></div>
                <div class="card ${netCls ? 'pos' : ''}"><div class="label" data-i18n="view.househack.card.net">Your net housing cost</div>
                    <div class="value ${netCls}">${money(r.net_housing_cost_usd)}/mo</div></div>
                <div class="card"><div class="label" data-i18n="view.househack.card.savings">Saved vs renting</div>
                    <div class="value ${saveCls}">${money(r.savings_vs_renting_usd)}/mo</div></div>
                <div class="card"><div class="label" data-i18n="view.househack.card.cashflow">Cash flow if you move out</div>
                    <div class="value ${cfCls}">${money(r.full_rental_cash_flow_usd)}/mo</div></div>
            </div>
            ${chart}
            <p class="muted small">${r.rent_covers_pi
                ? `<span class="pos" data-i18n="view.househack.note.covers">One unit's rent already covers the mortgage P&amp;I.</span>`
                : `<span data-i18n="view.househack.note.partial">The rented units cover part of the carrying cost — the rest is your housing budget.</span>`}</p>
        </div>
    `;
    applyUiI18n(el);
}
